src/array/cuda/labor_sampling.cu

```diff
@@ -0,0 +1,858 @@
+  std::unique_ptr<T, workspace_memory_alloc> alloc_unique(std::size_t size) {
+    return std::unique_ptr<T, workspace_memory_alloc>(reinterpret_cast<T *>(
+        runtime::DeviceAPI::Get(*ctx)->AllocWorkspace(*ctx, sizeof(T) * size)), *this);
+  }
+
+  char *allocate(std::ptrdiff_t size) {
+    return reinterpret_cast<char *>(runtime::DeviceAPI::Get(*ctx)->AllocWorkspace(*ctx, size));
+  }
+
+  void deallocate(char *ptr, std::size_t) {
+    runtime::DeviceAPI::Get(*ctx)->FreeWorkspace(*ctx, ptr);
+  }
+};
+
+template <typename IdType>
+struct TransformOp {
+  const IdType * idx_coo;
+  const IdType * rows;
+  const IdType * indptr;
+  const IdType * subindptr;
+  const IdType * indices;
+  const IdType * data_arr;
+  __host__ __device__
+  auto operator() (IdType idx) {
+    const auto in_row = idx_coo[idx];
+    const auto row = rows[in_row];
+    const auto in_idx = indptr[row] + idx - subindptr[in_row];
+    const auto u = indices[in_idx];
+    const auto data = data_arr ? data_arr[in_idx] : in_idx;
+    return thrust::make_tuple(row, u, data);
+  }
+};
+
+template <typename IdType, typename FloatType, typename probs_t, typename A_t, typename B_t>
+struct TransformOpImp {
+  probs_t probs;
+  A_t A;
+  B_t B;
+  const IdType * idx_coo;
+  const IdType * rows;
+  const FloatType * cs;
+  const IdType * indptr;
+  const IdType * subindptr;
+  const IdType * indices;
+  const IdType * data_arr;
+  __host__ __device__
+  auto operator() (IdType idx) {
+    const auto ps = probs[idx];
+    const auto in_row = idx_coo[idx];
+    const auto c = cs[in_row];
+    const auto row = rows[in_row];
+    const auto in_idx = indptr[row] + idx - subindptr[in_row];
+    const auto u = indices[in_idx];
+    const auto w = A[in_idx];
+    const auto w2 = B[in_idx];
+    const auto data = data_arr ? data_arr[in_idx] : in_idx;
+    return thrust::make_tuple(in_row, row, u, data, w / min((FloatType)1, c * w2 * ps));
+  }
+};
+
+template <typename FloatType>
+struct StencilOp {
+  const FloatType * cs;
+  template <typename IdType>
+  __host__ __device__
+  auto operator() (IdType in_row, FloatType ps, FloatType rnd) {
+    return rnd <= cs[in_row] * ps;
+  }
+};
+
+template <typename IdType, typename FloatType, typename ps_t, typename A_t>
+struct StencilOpFused {
+  const uint64_t rand_seed, rand_seed2;
+  const FloatType a, b;
+  const IdType * idx_coo;
+  const FloatType * cs;
+  const ps_t probs;
+  const A_t A;
+  const IdType * subindptr;
+  const IdType * rows;
+  const IdType * indptr;
+  const IdType * indices;
+  const IdType * nids;
+  __device__
+  auto operator() (IdType idx) {
+    const auto in_row = idx_coo[idx];
+    const auto ps = probs[idx];
+    IdType rofs = idx - subindptr[in_row];
+    const IdType row = rows[in_row];
+    const auto in_idx = indptr[row] + rofs;
+    const auto u = indices[in_idx];
+    const auto v = nids ? nids[u] : u;
+    curandStatePhilox4_32_10_t rng;
+    if (labor)
+      curand_init(123123, rand_seed, v, &rng);
```

Comment:
there is no reason for 123123, it was randomly chosen by me. I read from the documentation that if one wants statistically uncorrelated randoms, then same random seed should be used. That is why I pass 123123 as the seed. This way of using it uses 123123 as the seed and samples the 2^64*rand_seed + v th number in the PRNG sequence. So, the rolled random number is a function of rand_seed, vertex id and also all the rolled randoms are statistically uncorrelated.

Reply:
With different seeds, given an integer i, ith number from two different seeds has the potential to be correlated.

Reply:
Rolling randoms that depend on the vertex id and seed is an implementation detail that can be taken advantage of to get other functionalities. It also eliminates a sort or the use of a hashmap, so it is an optimization, to ensure a single random is rolled for the same vertex.

This seed logic is what makes labor sampling efficient, otherwise I couldn't have shown that it is faster than neighbor sampling.

Reply:
if labor is set to false, then labor sampling reduces to neighbor sampling, will put a comment referencing the paper.

Reply:
Why would rolling for each vertex ID individually remove the necessity of a hashmap?

Reply:
Because the random to be rolled, called r_t in the paper where t is the vertex id has to be a function of the random seed and the vertex id t. If I don't use this trick, then I have to take the unique operation on the next layer's potential vertices, roll the random numbers and then scatter the rolled randoms back to the edges.

Reply:
If the rolled random was called r_{ts} so that the rolled random for each edge is different, where t is source id, s is destination id when sampling in edges, then we get the same behaviour as neighbor sampling. setting the constexpr labor to false gets this behaviour and then this code would turn into neighbor sampling.

Reply:
One can set the constexpr labor variable to false to see how much slower this implementation is compared to the cuda neighbor sampling implementation when doing the same work, which is doing neighbor sampling.

Reply:
If you checkout the current version of the cpu implementation, you will see that we store the rolled randoms into a hashmap. Then we lookup random numbers from the hashmap while doing sampling. This trick avoids that kind of extra computation/communication(meaning memory access).

Reply:
NVIDIA documentation says the following: Sequences generated with different seeds usually do not have statistically correlated values, but some choices of seeds may give statistically correlated sequences.

Reply:
> If I don't use this trick, then I have to take the unique operation on the next layer's potential vertices, roll the random numbers and then scatter the rolled randoms back to the edges which is how I was doing things before I noticed this trick.

Does that mean I no longer need a unique operation if each seed node gets its own RNG?  If two seed nodes share some neighbors, it's still possible for two vertices to pick the same neighbor no?

Reply:
It is not that each seed node gets its own RNG, let's say we are iterating over an edge with seed node s and neighbor t. Since we init the rng with random seed and t, the rolled random for this edge will be the same as the rolled random for any other edge with seed node s' and neighbor t.